### ISSUE TITLE
CFS scheduler bug throttles highly threaded I/O blocked applications in Kubernetes

### ISSUE THREAD

**Issue by dharmab** [coreos/bugs#2623]
# Issue Report #

### Container Linux Version ###

```
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=2191.5.0
VERSION_ID=2191.5.0
BUILD_ID=2019-09-04-0357
PRETTY_NAME="Container Linux by CoreOS 2191.5.0 (Rhyolite)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"
```

### Environment ###

Azure, AWS, and VMware

### Expected Behavior ###

Highly threaded, I/O blocked containers running on Kubernetes on CoreOS should be able to use their full configured CPU request.

### Actual Behavior ###

Highly threaded, I/O blocked containers running on Kubernetes on CoreOS are heavily throttled well before they approach their CPU request. We are seeing **CPU performance impact of 50% in production for some Java web applications**- i.e. if we request 6 cores, we are throttled at 3.

See https://lkml.org/lkml/2019/5/17/581 for the kernel patch to fix this, which is landing in 5.4.

Because this bug heavily impacts the main intended use case of CoreOS Container Linux, would it be possible to prioritize this patch for backport?

**Comment by jcrowthe** [comment]
Small clarification: the issue lies in `cfs`, the CPU bandwidth control mechanism. Due to this bug in the kernel, `cfs` may throttle a pod well before its `requests` are reached. Hence the issue is in how the kernel enforces the kubernetes pod limits rather than pod requests.

**Comment by jhohertz** [comment]
Just wanted to reference some work I've done on this, overlay commit w/ the patch here:

https://github.com/viafoura/coreos-overlay/commit/b32f07fec474333afc8edf0bd9dc4ca8ae719d08

**Comment by chiluk** [comment]
cpu ".request" = cpu.shares which is a different mechanism from ".limits" which uses cfs bandwidth control.  If you really are using .request you are probably being hit by over-committed cpus on your nodes.

That being said yes CoreOS should backport the following linux commits onto their kernel if they haven't already.
512ac999
de53fd7ae
763a9ec06

**Comment by jhohertz** [comment]
Update: I am reworking my PR w/ input from the developers now.

**Comment by dharmab** [comment]
@chiluk good callout. We were setting both requests and limits to 6 in our test, and reproduced the throttling on a node we tainted to run just the one application Pod that had many more cores available.

@jhohertz thanks for submitting the patches! We've built a CoreOS image with those patches for some testing and are also awaiting an official release.

**Comment by dharmab** [comment]
We confirmed this patch fixed our throttling issue! Thank you!

**Comment by bgilbert** [comment]
This should be fixed in alpha 2317.0.1, due shortly.  We'll probably roll the fix into beta and stable more quickly than the normal promotion schedule would suggest.  Reopening until that happens.

**Comment by bgilbert** [comment]
This will also be fixed in beta 2303.2.0 and stable 2247.7.0, due shortly.  Thanks for reporting.
